QRCoder/QRCodeGenerator.cs

```diff
@@ -910,72 +989,81 @@ private static bool IsInRange(char c, char min, char max)
             return (uint)(c - min) <= (uint)(max - min);
         }
 
-        private static Polynom CalculateMessagePolynom(string bitString)
+        private static Polynom CalculateMessagePolynom(BitArray bitArray, int offset, int count)
         {
-            var messagePol = new Polynom();
-            for (var i = bitString.Length / 8 - 1; i >= 0; i--)
+            var messagePol = new Polynom(count / 8);
+            for (var i = count / 8 - 1; i >= 0; i--)
             {
-                messagePol.PolyItems.Add(new PolynomItem(BinToDec(bitString.Substring(0, 8)), i));
-                bitString = bitString.Remove(0, 8);
+                messagePol.PolyItems.Add(new PolynomItem(BinToDec(bitArray, offset, 8), i));
+                offset += 8;
             }
             return messagePol;
         }
 
 
         private static Polynom CalculateGeneratorPolynom(int numEccWords)
         {
-            var generatorPolynom = new Polynom();
-            generatorPolynom.PolyItems.AddRange(new[]{
-                new PolynomItem(0,1),
-                new PolynomItem(0,0)
-            });
+            var generatorPolynom = new Polynom(2);
+            generatorPolynom.PolyItems.Add(new PolynomItem(0, 1));
+            generatorPolynom.PolyItems.Add(new PolynomItem(0, 0));
+            var multiplierPolynom = new Polynom(numEccWords * 2);
             for (var i = 1; i <= numEccWords - 1; i++)
             {
-                var multiplierPolynom = new Polynom();
-                multiplierPolynom.PolyItems.AddRange(new[]{
-                   new PolynomItem(0,1),
-                new PolynomItem(i,0)
-                });
+                multiplierPolynom.PolyItems.Clear();
+                multiplierPolynom.PolyItems.Add(new PolynomItem(0, 1));
+                multiplierPolynom.PolyItems.Add(new PolynomItem(i, 0));
 
                 generatorPolynom = MultiplyAlphaPolynoms(generatorPolynom, multiplierPolynom);
             }
 
             return generatorPolynom;
         }
 
-        private static List<string> BinaryStringToBitBlockList(string bitString)
+        private static byte[] BinaryStringToBitBlockByteList(BitArray bitString, int offset, int count)
         {
             const int blockSize = 8;
-            var numberOfBlocks = (int)Math.Ceiling(bitString.Length / (double)blockSize);
-            var blocklist = new List<string>(numberOfBlocks);
-
-            for (int i = 0; i < bitString.Length; i += blockSize)
+            if (count % blockSize != 0)
+                throw new ArgumentException("Count must be a multiple of 8.", nameof(count));
+            var numberOfBlocks = count / blockSize;
```

Comment:
```suggestion
            var numberOfBlocks = (int)((uint)count / blockSize);
```
Then it's just a bit-operation instead of the integer-division.